File: 463. 整数排序.cpp
class Solution {
public:
    /**
     * @param A: an integer array
     * @return: nothing
     */
    int partition(vector<int>& v, int begin, int end) {
        int head = begin + 1;
        int tail = end;
        while (head < tail) {
            while (v[head] < v[begin] && head < tail) ++head;
            while (v[tail] > v[begin] && head < tail) --tail;
            if (head < tail) {
                swap(v[head], v[tail]);
                if (++head == tail || --tail == head) break;
            }
        }
        if (v[head] < v[begin]) {
            swap(v[head], v[begin]);
            return head;
        } else {
            swap(v[head - 1], v[begin]);
            return head - 1;
        }
    }

    void quick_sort(vector<int>& v, int begin, int end) {
        if (begin >= end) return;
        int pivot = partition(v, begin, end);
        quick_sort(v, begin, pivot - 1);
        quick_sort(v, pivot + 1, end);
    }
     
    void sortIntegers(vector<int> &A) {
        quick_sort(A, 0, A.size() - 1);
    }
};

// 总耗时 308 ms
// 您的提交打败了 61.60% 的提交!

File: Python3/463. 整数排序.py
def partition(A, begin, end) :
    head = begin + 1
    tail = end
    while head < tail :
        while A[head] < A[begin] and head < tail : head += 1
        while A[tail] > A[begin] and head < tail : tail -= 1
        if head < tail :
            A[head], A[tail] = A[tail], A[head]
            head += 1
            if head == tail : break
            tail -= 1
            if head == tail : break
    if A[head] < A[begin] :
        A[head], A[begin] = A[begin], A[head]
        return head
    else :
        A[head - 1], A[begin] = A[begin], A[head - 1]
        return head - 1

def quickSort(A, begin, end) :
    if begin >= end : return
    pivot = partition(A, begin, end)
    quickSort(A, begin, pivot - 1)
    quickSort(A, pivot + 1, end)

class Solution:
    """
    @param A: an integer array
    @return: nothing
    """
    def sortIntegers(self, A):
        quickSort(A, 0, len(A) - 1)

# 总耗时 1171 ms
# 您的提交打败了 89.71% 的提交!